fendermint/vm/interpreter/src/chain.rs

```diff
@@ -173,32 +211,55 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDown => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    if !provider.is_enabled() {
+                        return Err(anyhow!("toggle on ipc top down to process messages"));
```

Comment:
```suggestion
                        return Err(anyhow!("cannot execute IPC top-down message: parent provider disabled"));
```
Sorry, I found that message a confusing, so I assume someone less familiar with the code would do even more so.